File: leetcode/347.top-k-frequent-elements/cpp/topk.cpp
#include <vector>
#include <unordered_map>

class Solution
{
public:
    std::vector<int> topKFrequent(std::vector<int> &nums, int k)
    {
        std::unordered_map<int, int> frequencyByNumber;

        for (auto n : nums)
            frequencyByNumber[n]++;

        std::vector<std::pair<int, int>> numberFrequencies;

        for (const auto item : frequencyByNumber)
            numberFrequencies.push_back({item.first, item.second});

        std::sort(
            numberFrequencies.begin(),
            numberFrequencies.end(),
            [](const auto &a, const auto &b)
            {
                return a.second > b.second;
            });

        std::vector<int> res;

        for (int i = 0; i < k; ++i)
            res.push_back(numberFrequencies[i].first);

        return res;
    }
};

File: leetcode/347.top-k-frequent-elements/ts/topk.ts
const topKFrequent = (nums: number[], k: number): number[] => {
    const frequencyByNumber = new Map<number, number>()

    for (const n of nums) frequencyByNumber.set(n, (frequencyByNumber.get(n) ?? 0) + 1)

    const frequencyEntries: { k: number, c: number }[] = Array()

    for (const [k, c] of frequencyByNumber)
        frequencyEntries.push({ k: k, c: c })

    frequencyEntries.sort((a, b) => b.c - a.c)

    return frequencyEntries.slice(0, k).map(entry => entry.k)
};

export default topKFrequent
